GitUI/BranchTreePanel/RepoObjectsTree.Nodes.cs

```diff
@@ -156,60 +165,55 @@ private void UICommands_PostRepositoryChanged(object sender, GitUIPluginInterfac
             public bool IgnoreSelectionChangedEvent { get; set; }
             protected GitModule Module => UICommands.Module;
 
+            public void CancelReloadNodes()
+            {
+                _reloadCancellationTokenSequence.CancelCurrent();
+            }
+
             // Invoke from child class to reload nodes for the current Tree. Clears Nodes, invokes
             // input async function that should populate Nodes, then fills the tree view with its contents,
             // making sure to disable/enable the control.
-            protected void ReloadNodes(Func<CancellationToken, Task> loadNodesTask)
+            protected async Task ReloadNodesAsync(Func<CancellationToken, Task> loadNodesTask)
             {
-                ThreadHelper.ThrowIfNotOnUIThread();
-
-                ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
-                {
-                    var repoObjectTree = TreeViewNode.TreeView.Parent;
-
-                    try
-                    {
-                        var token = _reloadCancellationTokenSequence.Next();
+                await TaskScheduler.Default;
 
-                        repoObjectTree.Enabled = false;
+                Nodes.Clear();
```

Comment:
Nodes are accessed from both background and main thread. I think that `loadNodesTask` should return new nodes and then the `Nodes` should be populated with them in the main thread. Access to Nodes should be restricted to the main thread only (they are not thread-safe). Something along the lines of:

```csharp
                await TaskScheduler.Default;

                var token = _reloadCancellationTokenSequence.Next();
                var newNodes = await loadNodesTask(token);

                await TreeViewNode.TreeView.SwitchToMainThreadAsync(token);

                Nodes.Clear();
                Nodes.AddNodes(newNodes);
```

Reply:
@sharwell also suggested to call `token.ThrowIfCancellationRequested();`
see https://github.com/gitextensions/gitextensions/pull/6273#discussion_r259640694

Reply:
Thanks! If there is a difference between controlSwitchToMainThreadAsync. and ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(); then i believe it is better to stick with `token.ThrowIfCancellationRequested();` all the places.

Reply:
I think `control.SwitchToMainThreadAsync(token)` is a shorthand for `TH.JTF.SwitchToMainThreadAsync(token)`

Reply:
No, as per the comment you cited, @sharwell said that they have different semantics.
`TH.JTF.SwitchToMainThreadAsync(token)` - switches to the main thread if the cancellation token was not cancelled. The token validation is not performed on the main thread.

`control.SwitchToMainThreadAsync(token)` - switches to the main thread and on the main thread checks if the control is not disposed and if the toke is valid.

Reply:
> Nodes are accessed from both background and main thread. I think that `loadNodesTask` should return new nodes and then the `Nodes` should be populated with them in the main thread.

Yes, I'll do this.